x-pack/plugin/eql/src/main/antlr/EqlBase.g4

```diff
@@ -21,43 +18,52 @@ singleExpression
 statement
     : query (PIPE pipe)*
     ;
- 
+
 query
     : sequence
     | join
-    | condition
+    | eventQuery
+    ;
+withParams
+    : WITH namedParam (COMMA namedParam)*
     ;
-    
 sequence
-    : SEQUENCE (by=joinKeys)? (span)?
-      match+
-      (UNTIL match)?
+    : SEQUENCE  (by=joinKeys withParams? | withParams by=joinKeys?)?
+      sequenceTerm sequenceTerm+
+      (UNTIL sequenceTerm)?
     ;
 
 join
     : JOIN (by=joinKeys)?
-      match+
-      (UNTIL match)?
+      joinTerm joinTerm+
+      (UNTIL joinTerm)?
     ;
 
 pipe
     : kind=IDENTIFIER (booleanExpression (COMMA booleanExpression)*)?
     ;
 
+
+namedParam: key=IDENTIFIER EQ (expression | timeUnit);
+
 joinKeys
-    : BY qualifiedNames
-    ;
-    
-span
-    : WITH MAXSPAN EQ DIGIT_IDENTIFIER
+    : BY expression (COMMA expression)*
```

Comment:
`expression` is quite loose vs identifier as it considers `foo +1` , `1/2` etc as valid grammar and it's not.

Reply:
This is intentional, and does mirror the [existing grammar](https://github.com/endgameinc/eql/blob/master/eql/etc/eql.g).

It'll likely have performant implications on the stack, but existing implementations of EQL do support this.

Reply:
Languages are defined by their syntax - how parsing is done, whether it's grammar that leads to generation, done by hand, etc.. it's an implementation detail.

Considering we're looking at reimplementing the parsing in a different language, with a different type of parser (LL vs PEG until v0.7, LALR now in 0.8) I don't think fully mirroring the grammar is attainable nor should be a goal.
Further more in this case, the grammar is incorrect so why keep that ?
Between correctness and compatibility, we should always chose the former.

Reply:
> Languages are defined by their syntax

Yes I agree. Even more precisely, I would say that languages are defined by the inputs they accept. I'm not advocating for mirroring a PEG grammar in ANTLR, but I do think that we should strive the same inputs and provide the same semantics as before. We can make exceptions as needed, and fix things along the way, but they should be intentional and we should be aware.

> Further more in this case, the grammar is incorrect so why keep that ?

I think this is where we disagree. The previous grammar and implementations actually _used_ this functionality. It wasn't too loose on accident. There were a few cases where the underlying schema wasn't enough, for `by` to match the underlying events together and we have had to use something akin to this, for example:
```
sequence
   [process where ...] by some_field
   [security where ...] by someFunction(other_field)
```

I think that if we ever have to choose between correctness and compatibility, choosing correctness makes sense. But in this case, the design was intentional so I don't think these are in conflict. If we need to restrict this because it's not scalable within ES, we can do that. But that doesn't mean the previous version was incorrect, just because there weren't many examples that exercised arbitrary expressions for a join key.